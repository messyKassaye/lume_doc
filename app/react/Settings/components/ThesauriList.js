import { I18NLink, t } from 'app/I18N';
import { checkThesaurusCanBeDeleted, deleteThesaurus } from 'app/Thesauri/actions/thesaurisActions';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Icon } from 'UI';
import sortThesauri from '../utils/sortThesauri';

export class ThesauriList extends Component {
  getThesaurusSuggestionActions(thesaurus) {
    const showSuggestions =
      this.props.topicClassificationEnabled || thesaurus.enable_classification;
    return (
      showSuggestions && (
        <div className="vertical-line">
          {thesaurus.enable_classification && (
            <span className="thesaurus-suggestion-count">
              {thesaurus.suggestions ? thesaurus.suggestions.toLocaleString() : 'No'}&nbsp;
              {t('System', 'documents to be reviewed')}
            </span>
          )}
          <I18NLink
            to={`/settings/dictionaries/cockpit/${thesaurus._id}`}
            className="btn btn-default btn-xs"
          >
            <span>{t('System', 'Configure suggestions')}</span>
          </I18NLink>
        </div>
      )
    );
  }

  getThesaurusModifyActions(thesaurus) {
    return (
      <div>
        <I18NLink
          to={`/settings/dictionaries/edit/${thesaurus._id}`}
          className="btn btn-default btn-xs"
          confirmTitle={
            thesaurus.enable_classification ? 'Confirm edit suggestion-enabled Thesaurus' : ''
          }
          confirmMessage={
            thesaurus.enable_classification
              ? 'Uwazi suggests labels based on the current content of the document collection and its metadata. ' +
                'Editing this thesaurus, the content of the documents or other metadata can affect Uwaziâ€™s understanding of what to suggest.'
              : ''
          }
        >
          <Icon icon="pencil-alt" />
          &nbsp;
          <span>{t('System', 'Edit')}</span>
        </I18NLink>
        {'  '}
        <button
          onClick={this.deleteThesaurus.bind(this, thesaurus)}
          className="btn btn-danger btn-xs template-remove"
          type="button"
        >
          <Icon icon="trash-alt" />
          &nbsp;
          <span>{t('System', 'Delete')}</span>
        </button>
      </div>
    );
  }

  deleteThesaurus(thesaurus) {
    return this.props
      .checkThesaurusCanBeDeleted(thesaurus)
      .then(() => {
        this.context.confirm({
          accept: () => {
            this.props.deleteThesaurus(thesaurus);
          },
          title: `Confirm delete thesaurus: ${thesaurus.name}`,
          message: 'Are you sure you want to delete this thesaurus?',
        });
      })
      .catch(() => {
        this.context.confirm({
          accept: () => {},
          noCancel: true,
          title: `Cannot delete thesaurus: ${thesaurus.name}`,
          message: 'This thesaurus is being used in document types and cannot be deleted.',
        });
      });
  }

  thesaurusNode(thesaurus) {
    return (
      <tr key={thesaurus.name}>
        <th scope="row">{thesaurus.name}</th>
        <td>{this.getThesaurusSuggestionActions(thesaurus)}</td>
        <td>{this.getThesaurusModifyActions(thesaurus)}</td>
      </tr>
    );
  }

  render() {
    return (
      <div className="flex panel panel-default">
        <div className="panel-heading">{t('System', 'Thesauri')}</div>
        <div className="thesauri-list">
          <table>
            <thead>
              <tr>
                <th className="nameCol" scope="col" />
                <th scope="col" />
                <th scope="col" />
              </tr>
            </thead>
            <tbody>
              {sortThesauri(this.props.dictionaries.toJS()).map(thesaurus =>
                this.thesaurusNode(thesaurus)
              )}
            </tbody>
          </table>
        </div>
        <div className="settings-footer">
          <I18NLink to="/settings/dictionaries/new" className="btn btn-success">
            <Icon icon="plus" />
            <span className="btn-label">{t('System', 'Add thesaurus')}</span>
          </I18NLink>
        </div>
      </div>
    );
  }
}

ThesauriList.propTypes = {
  dictionaries: PropTypes.object,
  topicClassificationEnabled: PropTypes.bool,
  deleteThesaurus: PropTypes.func.isRequired,
  checkThesaurusCanBeDeleted: PropTypes.func.isRequired,
};

ThesauriList.contextTypes = {
  confirm: PropTypes.func,
};

export function mapStateToProps(state) {
  return {
    dictionaries: state.dictionaries,
    topicClassificationEnabled: (state.settings.collection.toJS().features || {})
      .topicClassification,
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      deleteThesaurus,
      checkThesaurusCanBeDeleted,
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(ThesauriList);
