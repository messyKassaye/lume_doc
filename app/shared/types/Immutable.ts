import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
