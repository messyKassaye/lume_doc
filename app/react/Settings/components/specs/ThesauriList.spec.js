import { shallow } from 'enzyme';
import Immutable from 'immutable';
import React from 'react';

import { ThesauriList } from '../ThesauriList';

describe('ThesaurisList', () => {
  let component;
  let props;
  let context;

  beforeEach(() => {
    props = {
      deleteThesaurus: jasmine.createSpy('deleteThesaurus').and.returnValue(Promise.resolve()),
      checkThesaurusCanBeDeleted: jasmine
        .createSpy('checkThesaurusCanBeDeleted')
        .and.returnValue(Promise.resolve()),
      topicClassificationEnabled: true,
      dictionaries: Immutable.fromJS([
        {
          _id: 'thesaurusUnderscoreId1',
          name: 'Continents',
          values: [
            {
              _id: 'valueUnderscoreId1',
              label: 'Africa',
              id: 'valueId1',
            },
            {
              _id: 'valueUnderscoreId2',
              label: 'Asia',
              id: 'valueId2',
            },
            {
              _id: 'valueUnderscoreId3',
              label: 'North America',
              id: 'valueId3',
            },
            {
              _id: 'valueUnderscoreId4',
              label: 'South America',
              id: 'valueId4',
            },
            {
              _id: 'valueUnderscoreId5',
              label: 'Oceania',
              id: 'valueId5',
            },
            {
              _id: 'valueUnderscoreId6',
              label: 'Europe',
              id: 'valueId6',
            },
          ],
        },
        {
          _id: 'thesaurusUnderscoreId2',
          name: 'Issues',
          enable_classification: true,
          suggestions: 3,
          values: [
            {
              _id: 'valueUnderscoreId1',
              label: 'Detention',
              id: 'valueId1',
            },
            {
              _id: 'valueUnderscoreId2',
              label: 'Elections',
              id: 'valueId2',
            },
            {
              _id: 'valueUnderscoreId3',
              label: 'Freedom of Movement',
              id: 'valueId3',
            },
          ],
        },
      ]),
    };

    context = {
      confirm: jasmine.createSpy('confirm'),
    };
  });

  const render = () => {
    component = shallow(<ThesauriList {...props} />, { context });
  };

  describe('render', () => {
    it('should match the snapshot', () => {
      render();
      expect(component).toMatchSnapshot();
    });

    it('should omit non-enable_classification thesauris if toggle if off', () => {
      props.topicClassificationEnabled = false;
      render();
      const renderedContexts = component.find('table');
      expect(renderedContexts.find('td').find('.vertical-line').length).toBe(1);
    });

    it('should render suggestions nodes when suggestions exist', () => {
      render();
      const renderedContexts = component.find('table');
      expect(renderedContexts.length).toBe(1);
      expect(renderedContexts.find({ scope: 'row' }).length).toBe(2);
      expect(renderedContexts.find('td').find('.thesaurus-suggestion-count').length).toBe(1);
      expect(renderedContexts.find('td').find('.vertical-line').length).toBe(2);
      expect(
        renderedContexts
          .find('td')
          .find('.vertical-line')
          .at(0)
          .containsMatchingElement(<span>Configure suggestions</span>)
      ).toBeTruthy();
      expect(
        renderedContexts
          .find('.thesaurus-suggestion-count')
          .contains(
            <span className="thesaurus-suggestion-count">{3}&nbsp;documents to be reviewed</span>
          )
      ).toBeTruthy();
    });
  });

  describe('classification', () => {
    it('should confirm  before deleting the thesaurus', done => {
      render();
      component
        .instance()
        .deleteThesaurus({ _id: 'thesaurusUnderscoreId2', name: 'Issues' })
        .then(() => {
          expect(context.confirm).toHaveBeenCalled();
          expect(props.checkThesaurusCanBeDeleted).toHaveBeenCalled();
          done();
        });
    });
  });
});
