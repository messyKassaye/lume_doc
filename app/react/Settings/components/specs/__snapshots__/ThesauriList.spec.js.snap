// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThesaurisList render should match the snapshot 1`] = `
<div
  className="flex panel panel-default"
>
  <div
    className="panel-heading"
  >
    Thesauri
  </div>
  <div
    className="thesauri-list"
  >
    <table>
      <thead>
        <tr>
          <th
            className="nameCol"
            scope="col"
          />
          <th
            scope="col"
          />
          <th
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr
          key="Continents"
        >
          <th
            scope="row"
          >
            Continents
          </th>
          <td>
            <div
              className="vertical-line"
            >
              <Connect(I18NLink)
                className="btn btn-default btn-xs"
                to="/settings/dictionaries/cockpit/thesaurusUnderscoreId1"
              >
                <span>
                  Configure suggestions
                </span>
              </Connect(I18NLink)>
            </div>
          </td>
          <td>
            <div>
              <Connect(I18NLink)
                className="btn btn-default btn-xs"
                confirmMessage=""
                confirmTitle=""
                to="/settings/dictionaries/edit/thesaurusUnderscoreId1"
              >
                <Connect(Icon)
                  icon="pencil-alt"
                />
                 
                <span>
                  Edit
                </span>
              </Connect(I18NLink)>
                
              <button
                className="btn btn-danger btn-xs template-remove"
                onClick={[Function]}
                type="button"
              >
                <Connect(Icon)
                  icon="trash-alt"
                />
                 
                <span>
                  Delete
                </span>
              </button>
            </div>
          </td>
        </tr>
        <tr
          key="Issues"
        >
          <th
            scope="row"
          >
            Issues
          </th>
          <td>
            <div
              className="vertical-line"
            >
              <span
                className="thesaurus-suggestion-count"
              >
                3
                 
                documents to be reviewed
              </span>
              <Connect(I18NLink)
                className="btn btn-default btn-xs"
                to="/settings/dictionaries/cockpit/thesaurusUnderscoreId2"
              >
                <span>
                  Configure suggestions
                </span>
              </Connect(I18NLink)>
            </div>
          </td>
          <td>
            <div>
              <Connect(I18NLink)
                className="btn btn-default btn-xs"
                confirmMessage="Uwazi suggests labels based on the current content of the document collection and its metadata. Editing this thesaurus, the content of the documents or other metadata can affect Uwazi’s understanding of what to suggest."
                confirmTitle="Confirm edit suggestion-enabled Thesaurus"
                to="/settings/dictionaries/edit/thesaurusUnderscoreId2"
              >
                <Connect(Icon)
                  icon="pencil-alt"
                />
                 
                <span>
                  Edit
                </span>
              </Connect(I18NLink)>
                
              <button
                className="btn btn-danger btn-xs template-remove"
                onClick={[Function]}
                type="button"
              >
                <Connect(Icon)
                  icon="trash-alt"
                />
                 
                <span>
                  Delete
                </span>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="settings-footer"
  >
    <Connect(I18NLink)
      className="btn btn-success"
      to="/settings/dictionaries/new"
    >
      <Connect(Icon)
        icon="plus"
      />
      <span
        className="btn-label"
      >
        Add thesaurus
      </span>
    </Connect(I18NLink)>
  </div>
</div>
`;
